### PR TITLE
feat(store): local search, installed mark, markdown README

### PR DESCRIPTION
**Pull request by yukimemi**
## Summary

- `/` now does **local incremental search** over `plugin_name + description + topics`, matching the list TUI convention; `n` / `N` jump between matches.
- **`S` key** preserves the previous GitHub API search (`topic:neovim-plugin <query>`) for fetching beyond the cached result set.
- **Fetch up to three pages (~300 plugins)** from the GitHub Search API with per-page resilience (partial failures keep the already-fetched results).
- Render the README pane through `tui-markdown` (without the heavy `highlight-code` feature), strip common HTML tags (`` becomes its `alt` text; `<a>`/`<div>`/`<br>` etc. are removed) and prepend repo topics as a `Topics: ...` line with a separator.
- **Installed mark**: plugins already present in `config.toml` get a green `✓` in the leftmost column. Pressing `Enter` on an installed plugin shows `already installed: ...` instead of re-adding it.
- **Topics column**: rightmost column shows up to three topics (`#lua #ui ...`) in dark gray.
- Help popup, footer, README.md and CLAUDE.md are updated for the new keys and legend.

## Test plan

- [x] `cargo test` — 202 passed, including new unit tests covering `/` local search, `n`/`N` wrap, `Esc` cancel, installed-case matching, and the `installed_full_name` URL normalization helper.
- [x] `cargo clippy --all-targets -- -D warnings` is clean.
- [x] `cargo fmt --all -- --check` is clean.
- [ ] Manual `rvpm store` run:
  - [ ] ~300 plugins on first launch, green `✓` on already-installed ones, topics visible in the rightmost column.
  - [ ] `/lua` incrementally filters; `n` / `N` jump between matches; `Esc` clears; `Enter` keeps the pattern for further `n`/`N`.
  - [ ] `S` prompts for an API query and replaces the list with the result.
  - [ ] `Enter` on an installed plugin shows the warning message; `Enter` on a new plugin adds it and flips the row to `✓` on return.
  - [ ] README pane shows headings/lists/code blocks styled by `tui-markdown`, and badge `` blocks are reduced to readable alt text.
  - [ ] `R` clears the cache and reloads 300 plugins.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Local incremental search (`/`) across name/description/topics with `n`/`N` navigation; `S` triggers GitHub API search
  * README preview rendered as GitHub-flavored Markdown with syntax highlighting; installed plugins show a leading ✓
  * Discovery expanded to fetch more results (up to ~300 across pages); `R` refreshes cache; `o`, `s`, `Tab`, `?` keybindings updated

* **Behavior Changes**
  * `Enter` skips already-installed plugins with a warning; `Esc` cancels active input, `q` quits

* **Documentation**
  * Updated store command docs and keybindings

* **Tests**
  * Added tests for search, README handling, URL normalization, and installed-state behavior

### CODE REVIEW

File: src/store_tui.rs [hunk 1]
```diff
@@ -305,25 +598,50 @@ impl StoreTuiState {
...
+                }
+            })
+            .unwrap_or_default();
+
+        // バッジや `<div>` 等の HTML タグは pulldown-cmark が HTML block として扱うので
+        // 生テキストが残る。最低限、行頭の `<img ...>` / `<a ...>` / `</a>` / `<br>` 等を
+        // 除去して見た目を整える。
+        let cleaned_body = strip_common_html(&readme_body);
+        let combined = format!("{}{}", topics_prefix, cleaned_body);
+        let rendered = tui_markdown::from_str(&combined);
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/store_tui.rs]: Done in 1ea8662 — moved the strip+format behind a cache (`StoreTuiState::readme_prepared`) rekeyed on `(full_name, content_len, loading)`. Only `tui_markdown::from_str` runs per draw now; HTML strip is a single pass when content changes.

File: src/store_tui.rs [hunk 2]
```diff
@@ -52,13 +66,125 @@ impl SortMode {
...
+                let _ = remainder;
+                break;
+            }
+            if let Some(rel_end) = s[start..].find('>') {
+                s.replace_range(start..start + rel_end + 1, "");
+                continue;
+            }
+        }
+        break;
+    }
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/store_tui.rs]: Good catch on all three — fixed in 1ea8662. `remove_tag_instances` / `extract_attr` are gone; `strip_common_html` is now a single-pass scanner that only slices at ASCII `<`/`>`, so the `to_lowercase`-index panic (Turkish `İ` et al.) and the false-positive `<abbr>`/`<a>` break are both eliminated. Tests added for Japanese text, emoji next to ``, `<abbr>` vs `<a>`, and multi-byte alt values.

File: src/main.rs [hunk 3]
```diff
@@ -2655,20 +2655,68 @@ fn build_plugin_scripts(
...
 
 /// `rvpm store` — GitHub からプラグインを検索して追加する TUI。
+/// Plugin URL を GitHub の `owner/repo` 形式 (小文字) に正規化する。
+/// GitHub の `full_name` は大文字小文字非依存なので lowercase で揃える。
+/// - `"owner/repo"` → `Some("owner/repo")`
+/// - `"https://github.com/Owner/Repo(.git)?"` → `Some("owner/repo")`
+/// - `"git@github.com:Owner/Repo.git"` → `Some("owner/repo")`
+/// - GitHub 以外 (gitlab 等) → None
+fn installed_full_name(url: &str) -> Option<String> {
+    let trimmed = url.trim().trim_end_matches(".git");
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/main.rs]: Done in 1ea8662 — `installed_full_name` now trims trailing `/` both before and after the `.git` strip, so `folke/snacks.nvim/`, `.../Repo.git/` and `.../Repo/` all normalise correctly. Covered by `test_installed_full_name_trailing_slash`.

File: src/store.rs [hunk 4]
```diff
@@ -110,25 +114,60 @@ pub fn search_plugins(cache_root: &Path, query: &str) -> Result<Vec<GitHubRepo>>
...
+            }
+        };
+
+        let parsed: SearchResponse = match resp.json() {
+            Ok(p) => p,
+            Err(e) => {
+                if all_items.is_empty() {
+                    return Err(e.into());
+                }
+                eprintln!("\u{26a0} GitHub search page {} parse failed: {}", page, e);
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/store.rs]: Done in 1ea8662 — `eprintln!` removed from `search_plugins` since it's called while the TUI holds the alternate screen. Partial pagination failures now only return the already-fetched items silently (and skip the 24 h cache write, per the other review comment).

File: src/store_tui.rs [hunk 5]
```diff
@@ -52,13 +66,125 @@ impl SortMode {
...
+        {
+            let tag = &s[i..=i + rel_end];
+            let alt = extract_attr(tag, "alt").unwrap_or_default();
+            if !alt.is_empty() {
+                out.push_str(&alt);
+            }
+            i += rel_end + 1;
+            continue;
+        }
+        out.push(bytes[i] as char);
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/store_tui.rs]: Fixed in 1ea8662 (same commit that rewrites `strip_common_html`). The scanner no longer touches raw bytes — it slices the input only at ASCII `<`/`>` boundaries and pushes whole string slices, so Japanese, emoji and accented text in README bodies survive intact. `test_strip_html_preserves_japanese_text` / `test_strip_html_preserves_emoji_around_img` guard it.

File: src/main.rs [hunk 6]
```diff
@@ -2655,20 +2655,68 @@ fn build_plugin_scripts(
...
+        let set: std::collections::HashSet<String> = config
+            .plugins
+            .iter()
+            .filter_map(|p| installed_full_name(&p.url))
+            .collect();
+        (cache, set)
     } else {
-        resolve_cache_root(None)
+        (resolve_cache_root(None), std::collections::HashSet::new())
     };
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/main.rs:2744]: Done in 1ea8662 — `run_store` reads and parses `config.toml` through a `'resolve:` labelled block with explicit match fallbacks. If the file is missing, unreadable, or fails `parse_config`, it warns to stderr (still safe: we haven't entered the alt screen yet) and continues with `resolve_cache_root(None)` + an empty installed set. Aligns with the project's resilience principle.

File: src/store_tui.rs [hunk 7]
```diff
@@ -52,13 +66,125 @@ impl SortMode {
...
+        i += 1;
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/store_tui.rs]: Covered by the same `strip_common_html` rewrite in 1ea8662 (see the reply on the other UTF-8 thread). The new scanner iterates by char-aligned byte ranges and copies whole `&str` slices rather than raw bytes.

File: src/store.rs [hunk 8]
```diff
@@ -110,25 +114,60 @@ pub fn search_plugins(cache_root: &Path, query: &str) -> Result<Vec<GitHubRepo>>
...
+        }
+    }
 
     // キャッシュに保存
     if let Some(parent) = cache_path.parent() {
         std::fs::create_dir_all(parent).ok();
     }
-    let json = serde_json::to_string(&resp.items)?;
+    let json = serde_json::to_string(&all_items)?;
     std::fs::write(&cache_path, json).ok();
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/store.rs]: Done in 1ea8662 — added a `cache_complete` flag that flips to `false` if any page's request or JSON parse fails. The cache is only written when every requested page succeeded (or we stop early because GitHub returned a partial page / empty page legitimately). Partial results are still returned to the caller so the TUI can display what we got.

File: src/main.rs [hunk 9]
```diff
@@ -2729,14 +2801,26 @@ async fn run_store() -> Result<()> {
...
+                    _ => {}
+                }
+                continue;
+            }
+
+            // `S` GitHub API 検索モード (旧 `/` の挙動)
+            if state.api_search_mode {
+                match key.code {
+                    crossterm::event::KeyCode::Esc => state.search_cancel(),
                     crossterm::event::KeyCode::Enter => {
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/main.rs:2840]: Fixed in 2fb0b69 — the Esc handler for `api_search_mode` no longer calls `search_cancel()` (which wipes both modes). It now only clears `api_search_mode` and `search_input`, so any pre-existing local `/` search pattern and matches are preserved and n/N keep working after cancelling an API query.

File: src/store_tui.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/store_tui.rs]: @gemini-code-assist Addressed in 1ea8662 — the HTML strip + topics prefix are now cached via `StoreTuiState::readme_prepared` keyed on `(full_name, content_len, loading)`, so `draw()` only pays `tui_markdown::from_str` per frame. Does this look sufficient to you?

File: src/store_tui.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/store_tui.rs]: @gemini-code-assist Fixed in 1ea8662 — `remove_tag_instances` and the dual `to_lowercase`-plus-byte-index approach are gone. `strip_common_html` is now a single-pass scanner that slices only at ASCII `<`/`>` boundaries, so the Turkish `İ`-style index panic and the `<abbr>` vs `<a>` false positive are both eliminated. Tests cover Japanese text, emoji, `<abbr>`/`<a>`, HTML comments, and multi-byte alt attributes. Please take another look.

File: src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/main.rs]: @gemini-code-assist Applied in 1ea8662 — `installed_full_name` now runs `.trim_end_matches('/')` both before and after stripping `.git`, so `folke/snacks.nvim/`, `.../Repo.git/` and `.../Repo/` all normalise correctly. Added `test_installed_full_name_trailing_slash` to guard it.

File: src/store.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/store.rs]: @gemini-code-assist Addressed in 1ea8662 — both `eprintln!` calls inside `search_plugins` are removed since the TUI owns the alternate screen while this function runs. Mid-pagination failures are now silent at the IO layer: the partial result is still returned to the caller, and `cache_complete` prevents the 24 h cache from being poisoned with incomplete data.

File: src/store_tui.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/store_tui.rs]: @gemini-code-assist Fixed in 1ea8662 — this line is gone. The rewritten `strip_common_html` no longer touches raw bytes; it iterates with char-aligned byte indices (only slicing at ASCII `<`/`>`) and pushes whole `&str` slices via `out.push_str(&input[i..lt])`. Japanese/emoji/accented content in READMEs is preserved, and `test_strip_html_preserves_japanese_text` / `test_strip_html_preserves_emoji_around_img` guard against regressions.

**Comment by yukimemi** [comment]
@gemini-code-assist @coderabbitai the branch has grown since the last review — please take another pass. Summary of changes on top of `83a43a8` (the commit you last reviewed):

## Runtime fixes

- **`2c208db`** enable `highlight-code` feature of tui-markdown for syntax-highlighted code blocks in the README preview.
- **`3f55285`** strip background colours from rendered spans so zellij doesn't carry stale coloured cells when scrolling.
- **`3118ade` / `085a7f2`** clear the README pane (and the whole frame) before every draw to prevent cell carryover.
- **`800bb96` / `4c6502b`** call `terminal.clear()` on plugin selection change and on README scroll so ratatui's previous-frame buffer can't retain styled cells.
- **`2fb0b69`** Esc during `S` (API search) now only cancels the API input instead of wiping the pending local `/` search pattern.

## Layout & text fixes

- **`d034240`** adaptive layout: stack list-over-README vertically when the terminal is narrower than 160 cols; flexible column constraints so plugin names stay visible.
- **`99a9305`** strip Private Use Area / variation selector chars from the list cells so nerd-font icons don't push columns right.
- **`78d9a2a`** apply the same sanitizer to the README body — fixes wrap drift when a README contains nerd-font chars.

## README scroll

- **`c8b4e92`** track `readme_line_count` / `readme_visible_height` and clamp `scroll_readme_down` / `scroll_readme_to_bottom` so G lands on the last content line instead of blank space.
- **`8798699` / `085a7f2`** compute the ceiling from tui-markdown's rendered `Text` + a wrap-aware row estimate using `unicode-width`; `visible_width` is part of the cache key so resize invalidates it.

## Table rendering

- **`feccfa9`** pre-process GFM pipe tables: detect header + `---` separator + body rows, realign column widths via `unicode-width`, wrap the whole thing in a ```text``` fence so tui-markdown doesn't collapse the rows into a single Line. Minimum-viable plaintext table rendering that works without an external tool.
- **`51a0235` / `db0db0e`** earlier realign-only attempt reverted because tui-markdown still treated the output as a table and collapsed the rows.

All commits on the branch build cleanly; 222 tests pass; `cargo clippy --all-targets -- -D warnings` is clean; `cargo fmt --all -- --check` is clean.

Would particularly appreciate scrutiny of:

- The new `wrap_tables_as_code_blocks` / `sanitize_cell_text` / `estimate_wrapped_rows` helpers in `src/store_tui.rs`.
- The scroll clamp flow (line-count updates in `ensure_readme_prepared`, visible dimensions captured in `draw`, `terminal.clear()` placement in `run_store`).
- Any edge cases in the pipe-table detection (I covered unit tests for header/separator/empty-cell/non-table lines).

Thanks!

File: src/main.rs [hunk 10]
```diff
@@ -2688,8 +2760,15 @@ async fn run_store() -> Result<()> {
...
     let mut last_selected: Option<String> = None;
+    // README pane の scroll が変化したら terminal.clear() して diff を無効化する。
+    // highlight-code の styled span が zellij 等で残骸を残す問題への belt-and-suspenders。
+    let mut last_readme_scroll: u16 = state.readme_scroll;
 
     loop {
+        if state.readme_scroll != last_readme_scroll {
+            terminal.clear()?;
+            last_readme_scroll = state.readme_scroll;
+        }
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/main.rs:2775]: @gemini-code-assist This one is intentional and empirically needed, despite the theoretical redundancy. We tried Clear widget at pane scope, then at `f.area()` scope (in `StoreTuiState::draw`) — on zellij the diff engine still leaves syntect background bands when the scroll position changes. The user reproduced it consistently. `terminal.clear()` here only fires when `state.readme_scroll` actually moves (not per frame), and it's paired with an immediate `terminal.draw`, so no visible flash on any terminal we've tested. Without it the README preview is unreadable in zellij. Open to a better approach if you know one, but for now this is the working trade-off.

File: src/main.rs [hunk 11]
```diff
@@ -2711,6 +2790,10 @@ async fn run_store() -> Result<()> {
                 state.readme_loading = true;
                 state.readme_content = None;
                 state.readme_scroll = 0;
+                // Clear widget だけでは ansi-to-tui の styled span の残骸が
+                // 一部ホスト (zellij 等) で残ることがあるため、選択変更時は
+                // ratatui の内部バッファを明示的に無効化して全セル再描画を強制する。
+                terminal.clear()?;
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/main.rs:2800]: @gemini-code-assist Same rationale as the scroll-clear thread above: the Clear widget in `draw` wasn't sufficient on zellij, so a `terminal.clear()` fires only on selection change (discrete event, not per frame) and is immediately followed by a fresh `terminal.draw`. The flash is imperceptible in testing; skipping it regresses to a broken alt-screen on the user's setup. If a future ratatui release tightens the diff-vs-terminal contract, this call will be the first thing to go.

File: src/store_tui.rs [hunk 12]
```diff
@@ -305,25 +863,31 @@ impl StoreTuiState {
...
+        let mut rendered = tui_markdown::from_str(&self.readme_prepared);
+        // highlight-code 由来の背景色付き Span が scroll 時に一部ホストで
+        // 残骸を残すので、前景色だけ残して背景は既定に戻す。fg による
+        // syntax highlighting は維持されるので可読性は保たれる。
+        for line in &mut rendered.lines {
+            for span in &mut line.spans {
+                span.style.bg = None;
+            }
+            line.style.bg = None;
+        }
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/store_tui.rs]: @gemini-code-assist Fixed in 2492252 — added `text_to_owned` that walks the tui-markdown result once and produces a `Text<'static>` with owned `String` content per Span. The heavy work (pulldown-cmark parse, syntect highlight, bg strip) now runs inside `ensure_readme_prepared` and the result is cached in `readme_rendered`, keyed on the same `(full_name, content_len, loading, visible_width)` tuple as `readme_prepared`. `draw()` only pays a `Text::clone` per frame now. Resize automatically invalidates via `visible_width` in the cache key.

File: src/store_tui.rs [hunk 13]
```diff
@@ -16,11 +17,43 @@ pub enum Focus {
...
+    /// GitHub API 検索の入力モード (`S` キー) — 旧 `/` の挙動を退避
+    pub api_search_mode: bool,
+    /// search_mode / api_search_mode で共有する入力バッファ
     pub search_input: String,
+    /// 確定済み検索パターン (n/N 用)
+    pub search_pattern: Option<String>,
+    /// 検索にヒットした plugins のインデックス一覧
+    pub search_matches: Vec<usize>,
+    /// search_matches 内の現在位置
+    pub search_cursor: usize,
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/store_tui.rs:31]: @coderabbitai Fixed in a31ed4f. `sort_plugins` now calls `run_local_search` with the current `search_pattern` at the end of the sort, so `n`/`N` stay aligned with the reordered list. Added `test_sort_rebuilds_search_matches` that searches `/nvim` under Stars order (matches idx [1, 2]), flips to Name, and verifies the matches become [0, 1] with `n` landing on the next expected plugin.

File: src/store_tui.rs [hunk 14]
```diff
@@ -52,16 +85,289 @@ impl SortMode {
...
+            out.push_str("```\n");
+            i = end;
+            continue;
+        }
+        out.push_str(line);
+        out.push('\n');
+        i += 1;
+    }
+    out
+}
```
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/store_tui.rs:256]: @coderabbitai Fixed in a31ed4f. `wrap_tables_as_code_blocks` now tracks fenced state by toggling on any line whose trimmed text starts with three backticks or three tildes; table detection only kicks in when `in_fence` is false. Added two regression tests: `test_wrap_tables_skips_inside_fenced_code_block` (pipe table inside ```markdown stays untouched) and `test_wrap_tables_still_wraps_tables_outside_fences` (a real table after a lua fence still gets wrapped).

File: src/main.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (yukimemi) [reply src/main.rs:2775]: @gemini-code-assist Good hypothesis, but we tested exactly this sequence before landing 4c6502b. The order was: add bg strip (3f55285) → user reported residue still visible → add `terminal.clear()` on scroll (4c6502b) → residue gone. So the background-color strip is necessary but not sufficient on zellij; the diff engine apparently still holds on to styled cells even after their bg has been reset. Both safeguards are empirically needed to keep the pane artifact-free. If a future tui-markdown / ansi-to-tui / zellij combination tightens the contract, this `terminal.clear()` call is the first line we'd drop.
